### PR TITLE
ML-KEM documentation

### PR DESCRIPTION
**Pull request by baentsch**
Fixes https://github.com/openssl/private/issues/649.

Open question: Objections to putting all documentation for all ML-KEM algorithms in one file? References to other variants should be simply added to the 2 new core files doc/man7/EVP_KEM-MLKEM.pod and as and doc/man7/EVP_PKEY-MLKEM.pod when https://github.com/openssl/openssl/issues/26006 is resolved.

One new ToDo created referenced in this PR: https://github.com/openssl/openssl/issues/26036

### CODE REVIEW

**Comment by t8m** [comment]
> Open question: Objections to putting all documentation for all ML-KEM algorithms in one file?

No objections. IMO this is the right way. It is similar to the ECX algorithms.

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (mattcaswell) [inline doc/man7/EVP_KEM-MLKEM.pod:6]: IMO, this should look like this:
```suggestion
EVP_KEM-ML-KEM-768
```

The name after the `EVP_KEM-` prefix should be the name as it is fetched from the provider, i.e. the names as listed in the `PROV_NAMES_MLKEM768` define. It is not possible to "fetch" an algorithm called "ML-KEM" or "MLKEM" so we should not list them here. We can extend the list to include "ML-KEM-512" and "ML-KEM-768" as we implement them

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (mattcaswell) [inline doc/man7/EVP_PKEY-MLKEM.pod:8]: Again the only valid algorithm name here is `ML-KEM-768`. You can't "fetch" anything else.

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (mattcaswell) [inline doc/man7/OSSL_PROVIDER-default.pod:226]:
```suggestion
=item ML-KEM-768, see L<EVP_KEM-ML-KEM-768(7)>
```

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (mattcaswell) [inline doc/man7/OSSL_PROVIDER-default.pod:270]:
```suggestion
=item ML-KEM-768, see L<EVP_KEYMGMT-ML-KEM-768(7)>
```

File: doc/man7/provider-keymgmt.pod
Comment (mattcaswell) [inline doc/man7/provider-keymgmt.pod:505]:
```suggestion
L<EVP_PKEY-DSA(7)>, L<EVP_PKEY-DH(7)>, L<EVP_PKEY-ML-KEM-768(7)>
```

File: doc/man3/EVP_PKEY_decapsulate.pod
Comment (mattcaswell) [inline doc/man3/EVP_PKEY_decapsulate.pod:96]:
```suggestion
L<EVP_KEM-RSA(7)>, L<EVP_KEM-X25519(7)>, L<EVP_KEM-EC(7)>, L<EVP_KEM-ML-KEM-768(7)>
```

File: doc/man3/EVP_PKEY_encapsulate.pod
Comment (mattcaswell) [inline doc/man3/EVP_PKEY_encapsulate.pod:105]:
```suggestion
L<EVP_KEM-RSA(7)>, L<EVP_KEM-X25519(7)>, L<EVP_KEM-EC(7)>, L<EVP_KEM-ML-KEM-768(7)>
```

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (baentsch) [reply doc/man7/EVP_PKEY-MLKEM.pod:8]: Well, the name "EVP_PKEY-MLKEM" I only added after a test complained that it needs adding. The next two to allow the links from default provider to work -- exactly to avoid having to change more than 2 files as and when -512 and -1024 get added as per my initial comment.

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (baentsch) [reply doc/man7/OSSL_PROVIDER-default.pod:226]: OK -- so basically the statement is that all these files need updating as -512 and -1024 get added?

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (baentsch) [reply doc/man7/EVP_KEM-MLKEM.pod:6]: ACK. See below. Let me play with the doc-nits script how to bring your logical argument (that I agree with) and that check in line...

Comment (baentsch) [reply doc/man7/EVP_KEM-MLKEM.pod]: Now changed adding all possibly key types -- but had to leave "generic" EVP_KEM-MLKEM to satisfy `make doc-nits`.

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (baentsch) [reply doc/man7/EVP_PKEY-MLKEM.pod]: Changed equivalent as for PKEY above.

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (baentsch) [reply doc/man7/OSSL_PROVIDER-default.pod]: Changed to all specific key types in https://github.com/openssl/openssl/pull/26037/commits/5b7aad109796393b11b35d43474b444283ccb698

File: doc/man7/provider-keymgmt.pod
Comment (baentsch) [reply doc/man7/provider-keymgmt.pod]: Changed to all specific key types in https://github.com/openssl/openssl/pull/26037/commits/5b7aad109796393b11b35d43474b444283ccb698

File: doc/man3/EVP_PKEY_decapsulate.pod
Comment (baentsch) [reply doc/man3/EVP_PKEY_decapsulate.pod]: Changed to all specific key types in https://github.com/openssl/openssl/pull/26037/commits/5b7aad109796393b11b35d43474b444283ccb698

File: doc/man3/EVP_PKEY_encapsulate.pod
Comment (baentsch) [reply doc/man3/EVP_PKEY_encapsulate.pod]: Changed to all specific key types in https://github.com/openssl/openssl/pull/26037/commits/5b7aad109796393b11b35d43474b444283ccb698

File: doc/designs/mlkem.md
Comment (paulidale) [inline doc/designs/mlkem.md:75]: Over in #25990 I'm changing this parameter to `OSSL_KEM_PARAM_IKME` (_"ikme"_) so it aligns with the similar option used by the ECX algorithms.

It's not clear if it would be better to change it here too (before either merges).

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (paulidale) [inline doc/man7/EVP_KEM-MLKEM.pod:18]: `OSSL_KEM_PARAM_IKME` (_"ikme"_)  here too.

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (vdukhovni) [reply doc/man7/EVP_PKEY-MLKEM.pod:46]: Mostly gettable. What's settable are private or public keys, the seed for key generation and the entropy for encapsulation.

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (vdukhovni) [inline doc/man7/EVP_PKEY-MLKEM.pod:52]: This is its standard wire form encoding as the vector `t` and the matrix seed `rho`.

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (vdukhovni) [inline doc/man7/EVP_PKEY-MLKEM.pod:61]: Setting the public key clears any extant private key!

File: doc/designs/mlkem.md
Comment (baentsch) [reply doc/designs/mlkem.md:75]: Thanks for the head-up.

> It's not clear if it would be better to change it here too (before either merges).

My proposal would be to adapt this PR to whatever #25990 winds up encoding and merge it only at the end (when consensus has built what makes most sense/is most coherent with the pre-existing code base).

Comment (baentsch) [reply doc/designs/mlkem.md:75]: Thanks for the heads-up (and letting me know about #25990 :)!

> It's not clear if it would be better to change it here too (before either merges).

My thinking was to adapt this PR to whatever #26054 winds up encoding and merge it only at the end (when consensus has built what makes most sense/is most coherent with the new and the pre-existing code base). But I'd also be fine closing this PR if either you or @vdukhovni also want to add the documentation to #26054 straight.

Comment (paulidale) [reply doc/designs/mlkem.md:75]: I've pulled my changes out into #26082 which is likely to merge quickly.
I also fixed some bugs in the new evp_test KEM code.

**Comment by vdukhovni** [comment]
> I've pulled my IKME changes into a separate PR (#26082).

Reviewed, looks OK, but please rebase. The provider changes will otherwise conflict.

File: doc/designs/mlkem.md
Comment (baentsch) [reply doc/designs/mlkem.md:75]: FYI @paulidale we decided to merge this pretty much as-is to close out a sprint and then change docs again as new code lands.

File: doc/man7/OSSL_PROVIDER-default.pod
Comment (baentsch) [reply doc/man7/OSSL_PROVIDER-default.pod]: update done

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (baentsch) [reply doc/man7/EVP_PKEY-MLKEM.pod:62]: If I got the discussion right this morning, you'll be adding logic that needs to be described here anyway, so I leave "generic" as-is for now such as to close out this PR.

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (baentsch) [reply doc/man7/EVP_PKEY-MLKEM.pod]: fixed in latest commit.

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (baentsch) [reply doc/man7/EVP_KEM-MLKEM.pod:22]: This suggestion also seems like it is WIP. So leave wording as-is for now to close out the PR.

**Comment by baentsch** [comment]
@vdukhovni @mattcaswell I went through all feedback I felt possible to act upon now, commented on those that are still WIP (i.e., topic for a new iteration of documentation as agreed-upon this morning) and thus would like to suggest another review/merge for now so @vavroch2010 can clean out the sprint.

File: doc/designs/mlkem.md
Comment (mattcaswell) [inline doc/designs/mlkem.md:73]:
```suggestion
We naturally use this for running the Known Answer Tests, but our private key encoding
```

File: doc/designs/mlkem.md
Comment (mattcaswell) [inline doc/designs/mlkem.md:90]:
```suggestion
The "seed format" can be selected by setting the OSSL_PARAM value
```

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (mattcaswell) [inline doc/man7/EVP_KEM-MLKEM.pod:18]: We should specify whether this is gettable or settable

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (mattcaswell) [reply doc/man7/EVP_PKEY-MLKEM.pod:44]: We should whether they are gettable/settable in the documentation

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Mostly good. Some minor tweaks to address and I could approve this.

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (vdukhovni) [reply doc/man7/EVP_KEM-MLKEM.pod:18]: Settable only.

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (vdukhovni) [reply doc/man7/EVP_KEM-MLKEM.pod:22]: Encapsulation normally consumes 32 bytes of RNG *entropy*, which is hashed to produce other random values within the encapsulation process.  Specifying this explicitly, instead of generating it randomly, make encasulation fully deterministic.

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (vdukhovni) [reply doc/man7/EVP_KEM-MLKEM.pod:34]: Please fix the above.

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (vdukhovni) [reply doc/man7/EVP_PKEY-MLKEM.pod:59]: Actually, at this point setting the public key is not possible if any key is already set, it can only be set for a key that is still under construction.

**Review by vdukhovni — CHANGES_REQUESTED** [review]

**Comment by baentsch** [comment]
@vdukhovni I see the merit of some of your change requests (some had already been incorporated prior to your feedback) in light of all your improvements. However, I understood our agreement from last week to close this out based on the then-current code base and for @vavroch2010 to create a new tracking issue for another sprint and a consequent new documentation PR after the dust has settled (after #26054 has landed) as the stated goal/limitation was to not shift this PR into another sprint.

I now see two mutually exclusive ways forward:

1) If the primary goal is to satisfy the "sprint process", then please adhere to the agreement then made (review to merge without taking into consideration #26054). If that'd be OK for you @vdukhovni then I'd think some of your latest PR feedback isn't applicable (yet) and this PR is ready to merge.
2) If the primary goal is to create correct documentation, then let's switch to my proposal from last week and keep this PR open until #26054 has landed. That then would be an ask to @vavroch2010 to change tack.

I'll wait with further work until a decision has been made on whether 1) or 2) "rules". _My_ preference is 2) ("process should not trump progress") but it is not what has been agreed last week.

**Comment by mattcaswell** [comment]
> Thanks @mattcaswell . All comments incorporated in latest commit.

@baentsch - I don't see my comments about specifying all params as either gettable or settable as having been addressed?

**Review by vdukhovni — APPROVED** [review]

**Comment by baentsch** [comment]
> > Thanks @mattcaswell . All comments incorporated in latest commit.
> 
> @baentsch - I don't see my comments about specifying all params as either gettable or settable as having been addressed?

@mattcaswell I am of the opinion that it has (or that no further documentation is warranted: Or do you also expect params to be documented that are default ones, i.e., OSSL_PKEY_PARAM_PUB_KEY, OSSL_PKEY_PARAM_PRIV_KEY, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, OSSL_PKEY_PARAM_BITS, OSSL_PKEY_PARAM_SECURITY_BITS, OSSL_PKEY_PARAM_MAX_SIZE ?

The one MLKEM PKEY specific one, "OSSL_PKEY_PARAM_MLKEM_SEED" is documented along this line and reads

> This optional parameter can be used to set the value prior to key pair generation.

Or do you prefer a specific additional statement along the lines "This parameter can only be set." or even (additionally?) "This parameter cannot be retrieved."? 

The MLKEM KEM "OSSL_KEM_PARAM_MLKEM_ENC_ENTROPY" param does have such very explicit statement even though upon re-reading that seems a bit "too much"/"in your face" in my eyes.

Just let me know what you prefer and/or is the norm for other docs and I'll adapt.

**Comment by mattcaswell** [comment]
> Or do you prefer a specific additional statement along the lines "This parameter can only be set." 

Something like that would be helpful I think. It is implied that it can only be set, but only by absence of anything being said about getting.

Nothing is said about the getability/setability of OSSL_PKEY_PARAM_PUB_KEY, OSSL_PKEY_PARAM_PRIV_KEY - so I think we should be more explicit here.

**Comment by vdukhovni** [comment]
I have a bunch of minor tweaks to existing docs with mostly breadcrumb comments about specific places to mention MLKEM in the future, and a couple of places where it was already possible before this PR lands.

Should I hold on to these, or push them to this PR?

**Comment by baentsch** [comment]
> > Or do you prefer a specific additional statement along the lines "This parameter can only be set."
> 
> Something like that would be helpful I think. It is implied that it can only be set, but only by absence of anything being said about getting.

Explicit statement added.
> 
> Nothing is said about the getability/setability of OSSL_PKEY_PARAM_PUB_KEY, OSSL_PKEY_PARAM_PRIV_KEY - so I think we should be more explicit here.

Usage reference added.

**Review by mattcaswell — APPROVED** [review]
@vdukhovni please reconfirm

File: doc/designs/mlkem.md
Comment (vdukhovni) [reply doc/designs/mlkem.md:106]: OK, we'll need to remember to do that. The parameter name will in fact soon be: `OSSL_KEM_PARAM_IKME`.
Here, I'd have been tempted to just get it over with, and replace the name with that.

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (vdukhovni) [inline doc/man7/EVP_KEM-MLKEM.pod:10]: Particularly, given the section title immediately below...
```suggestion
The B<ML-KEM> keytypes and parameters are described in L<EVP_PKEY-ML-KEM(7)>.
```

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (vdukhovni) [inline doc/man7/EVP_PKEY-MLKEM.pod:57]: Settable only once, if the key is currently blank. Once either a public or private key has been loaded, the keys become immutable and no further updates to change the value or type (public vs. private) are allowed.

**Review by vdukhovni — APPROVED** [review]

**Comment by baentsch** [comment]
Thanks @mattcaswell . All squashed, removed private email address. Ready to merge from my perspective

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (baentsch) [reply doc/man7/EVP_KEM-MLKEM.pod:10]: We'll probably have to re-do it all as and when there's a final decision on how to name things. And given that, we'll re-visit the documentation as a whole when we're getting closer to merge-to-master -- which we're imo still pretty far away from (PR feedback, encoding/decoding, seeing that hybrids work, etc.)

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (levitte) [reply doc/man7/EVP_PKEY-MLKEM.pod]: You could try adding this somewhere:

```

=for openssl generic

```

(the empty lines surrounding it are significant)

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (levitte) [reply doc/man7/EVP_KEM-MLKEM.pod:10]: Well...  this algorithm *is* commonly referred to as `ML-KEM` (for example: https://openquantumsafe.org/liboqs/algorithms/kem/ml-kem.html).  I for one don't see why we would need to create another variant of the name (i.e. without the dash).

File: doc/man7/EVP_PKEY-MLKEM.pod
Comment (levitte) [inline doc/man7/EVP_PKEY-MLKEM.pod:18]: POD acts in paragraph mode, so you *must* have an empty line before the `=for` line as well for it to function as intended.

Comment (baentsch) [reply doc/man7/EVP_PKEY-MLKEM.pod]: thx. corrected.

File: doc/man7/EVP_KEM-MLKEM.pod
Comment (baentsch) [reply doc/man7/EVP_KEM-MLKEM.pod]: That discussion happens elsewhere. IETF uses no-dashes, users probably would prefer it, but NIST defined dashes. The latest commit streamlined everything over to the dashed version but I'll not change again. Hopefully this is now finally good to merge. @vdukhovni fyi as I had to touch everything, I also already updated the entropy param to "ikme" so this is in line with your PR (to hopefully also land soon now).

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t-j-h** [comment]
You need to apply the MLKEM -> ML_KEM change in terms of the naming - which has happened already in the implementation PR.

File: doc/designs/ML-KEM.md
Comment (paulidale) [inline doc/designs/ML-KEM.md:28]: Type names are supposed to be uppercase.  I know this represents the current code but that code needs to change IMO.

Comment (baentsch) [reply doc/designs/ML-KEM.md:28]: I also know that :-/

My intention (I thought in meetings with @vavroch2010 and @mattcaswell agreed-upon to close out a sprint) was to finish off/clean up this documentation task with a separate PR _after_ the code is finalized (https://github.com/openssl/openssl/pull/26054 has landed) and all "naming issues" have been resolved in the code. With his latest feedback @t-j-h however demanded this PR to be in line with the WIP which I also tried to satisfy now (invalidating the reviews and changing casing once again, I know). Trying to satisfy everyone is a bit challenging when reviews straddle (way) more than 1 sprint.

Comment (vdukhovni) [reply doc/designs/ML-KEM.md:28]: Type names used in *PUBLIC* interfaces, are supposed to be in upper case. This is NOT a public interface. There are a few other internal type names that are not upper case. Admittedly, rare beasts:
```
include/crypto/ecx.h-} ecx_key_op_t;
include/crypto/evp.h-} evp_cipher_aead_asn1_params;
include/crypto/ml_kem.h-} ossl_ml_kem_vinfo;
include/crypto/ml_kem.h-} ossl_ml_kem_scalar;
include/internal/dane.h-} danetls_record;
include/internal/hashtable.h-} keyname;
include/internal/ktls.h:typedef struct tls_enable ktls_crypto_info_t;
include/internal/ktls.h:typedef struct tls_crypto_info_all ktls_crypto_info_t;
```
and some public ones:
```
include/openssl/ct.h.in-} ct_log_entry_type_t;
include/openssl/ct.h.in-} sct_version_t;
include/openssl/ct.h.in-} sct_source_t;
include/openssl/ct.h.in-} sct_validation_status_t;
include/openssl/ec.h-} point_conversion_form_t;
include/openssl/engine.h-} dynamic_MEM_fns;
include/openssl/engine.h-} dynamic_fns;
```

Anyway the `ctx` in question is about to be consolidated into the key, which is being renamed `ML_KEM_KEY`. The only question is whether the current state of the PR is sufficient as a phase-I working draft, to make it possible to start building hybrids, ..., that just need EVP support, and don't care about the internals.

Or, whether I should just push the cleaned up phase-II code for review.

**Review by vdukhovni — APPROVED** [review]

**Comment by t8m** [comment]
Squashed and merged to the feature branch. Thank you.

**Pull request closed without merging**
